### PR TITLE
Vladk/phase4

### PR DESCRIPTION
**Pull request by vlad-karnauch**
- Support for stylable
- broken down into component folders: `components/item`, `components/menu`, etc...

### CODE REVIEW

File: src/components/option/option.st.css [hunk 1]
```diff
@@ -0,0 +1,15 @@
+@namespace "Option";
+@st-import Item from '../item/item.st.css';
+
+.root{
+    -st-mixin: Item;
+    -st-states: disabled;
+}
+
+.root:hover {
+    cursor: pointer;
```
Comment (eladdo92) [inline src/components/option/option.st.css:10]: can be directly on `.root`

File: src/components/option/option.tsx [hunk 2]
```diff
@@ -12,11 +13,10 @@ interface TextOptionProps {
  * @param onClick
  * @param disabled
  */
-const TextOption: FC<TextOptionProps> = ({ text, onClick, disabled }) => {
-    const className = 'item ' + (disabled ? 'disabled-option disabled' : 'option');
+const Option: FC<OptionProps> = ({ text, onClick, disabled }) => {
```
Comment (PeterShershov) [inline src/components/option/option.tsx:16]: should be named differently to avoid confusion

File: src/components/option/option.tsx [hunk 3]
```diff
@@ -12,11 +13,10 @@ interface TextOptionProps {
...
     const maybeOnClick = useOnClick(text, onClick, disabled);
     return (
-        <li className={className} onClick={maybeOnClick}>
+        <li className={st(classes.root, { disabled: !!disabled })} onClick={maybeOnClick}>
```
Comment (PeterShershov) [inline src/components/option/option.tsx:19]: you can go with `disabled = false` as default in the arguments and pass `{ disabled }` to the element

File: src/components/option/option.tsx [hunk 4]
```diff
@@ -35,8 +35,8 @@ function useOnClick(
     return useCallback(maybeOnClick, [text, onClick, disabled]);
 }
 
-export default memo(TextOption, areEqual);
+export default memo(Option, areEqual);
 
-function areEqual(a: TextOptionProps, b: TextOptionProps): boolean {
+function areEqual(a: OptionProps, b: OptionProps): boolean {
```
Comment (PeterShershov) [inline src/components/option/option.tsx:40]: davka TextOption was better IMO

File: src/components/themed-dropdown.st.css [hunk 5]
```diff
@@ -0,0 +1,65 @@
+@namespace 'Project';
+
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:2]: Why "Project"?
also extra space on line 3

File: src/components/themed-dropdown.st.css [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    surface-color: white;
+    surface-color-secondary: #bbb;
+    color-primary: #03a9f4;
+}
+
+.themedDropdown {
+    --st-extends: Dropdown;
+}
+
+.themedDropdown Trigger {
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:22]: override Trigger using psuedo-classes:
`.themedDropdown::trigger`

File: src/components/themed-dropdown.st.css [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    overflow: auto; /* scrollbar */
+}
+
+.themedMenu:disabled {
+    border: 1px solid value(disabled-color);
+}
+
+.themedMenu Option:hover {
+    background: value(color-primary);
+    color: white;
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:55]: `themedMenu::option`

File: test-kit/dropdown/dropdown-driver.ts [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+
+    getTriggerText(): string | null {
+        return this.trigger.getText();
+    }
+
+    isDisabled(): boolean {
+        return this.menu.isDisabled() && this.trigger.isDisabled();
+    }
+
+    clickItem(i: number): void {
```
Comment (PeterShershov) [inline test-kit/dropdown/dropdown-driver.ts:35]: `clickItemByIndex`

File: test-kit/menu/menu-driver.ts [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        if (root === null) {
+            throw new Error('no menu');
+        }
+        return root;
+    }
+
+    isDisabled(): boolean {
+        const root = this.getRoot();
+        return root.matches(selectors.disabledState);
+    }
```
Comment (PeterShershov) [inline test-kit/menu/menu-driver.ts:26]: add JSDoc to all public methods, also in other drivers

File: test-kit/menu/menu-driver.ts [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+     *
+     * @param i option position in the container (0 - 1st, 1 - 2nd, etc...)
+     * @returns a driver to manipulate, and gauge status of the option
+     */
+    getItemDriver(i: number): ItemDriver {
+        const item = this.getItem(i);
+        return new ItemDriver(item);
+    }
+
+    private getItem(i: number) {
```
Comment (PeterShershov) [inline test-kit/menu/menu-driver.ts:47]: `...byIndex`

File: test/menu.spec.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+function makeDriver(optionItemsContainer: JSX.Element): MenuDriver {
+    const { container } = render(optionItemsContainer);
+    const driver = new MenuDriver(container);
+    return driver;
+}
+
+describe('Options Container', () => {
+    describe('open', () => {
+        it('opens when open', () => {
```
Comment (PeterShershov) [inline test/menu.spec.tsx:46]: clarify this description

**Review by PeterShershov — CHANGES_REQUESTED** [review]

File: src/components/divider/divider.tsx [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+import { style, classes } from './divider.st.css';
+
+/**
+ * React component implementing a divider between text options in the dropdown
+ */
+const Divider: FC = () => {
+    return <li className={style(classes.root)} />;
+};
+
+export default memo(Divider);
```
[the remark this answers is not in the document]
Comment (vlad-karnauch) [reply src/components/divider/divider.tsx:11]: I feel that this version is more readable: here is a component, and let's do it memo...

File: src/components/dropdown/dropdown.tsx [hunk 13]
```diff
@@ -16,24 +17,36 @@ import { createPortal } from 'react-dom';
...
     const [position, triggerRef] = usePosition();
 
     return (
-        <div className="dropdown" ref={triggerRef}>
+        <div className={st(classes.root, className)} ref={triggerRef}>
             <Trigger text={value} disabled={disabled} onClick={handleToggle} />
             {createPortal(
-                <OptionsContainer
+                <Menu
+                    className={menuClassName}
```
[the remark this answers is not in the document]
Comment (vlad-karnauch) [reply src/components/dropdown/dropdown.tsx:47]: yes we can

File: src/components/item/item-props.tsx [hunk 14]
```diff
@@ -6,46 +6,46 @@
...
     return {
-        type: OptionType.TEXT,
+        type: ItemType.TEXT,
         value: text,
         disabled: disabled,
     };
 }
 
-export function dividerOptionProps(): DividerOptionProps {
+export function dividerProps(): DividerProps {
```
[the remark this answers is not in the document]
Comment (vlad-karnauch) [reply src/components/item/item-props.tsx:47]: to match the previous function: 
I have `optionProps(...)`, and `dividerProps()` - each returns a different kind of "item"

File: src/components/option/option.tsx (hunk 3 above)
Comment (vlad-karnauch) [reply src/components/option/option.tsx:19]: very nice

File: test-kit/menu/menu-driver.ts (hunk 9 above)
Comment (vlad-karnauch) [reply test-kit/menu/menu-driver.ts:26]: Some of the stuff has comments, but the majority is really self explanatory: 
isOpen(), isDisabled()...
That's why, as you suggested, clickItemByIndex is better than clickItem(i)

File: src/components/item/item-props.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (PeterShershov) [reply src/components/item/item-props.tsx:47]: You can have both types of props providers - a function and a constant. It's creating a new instance and runs an operation for no reason.

File: src/components/themed-dropdown.st.css [hunk 15]
```diff
@@ -0,0 +1,62 @@
+@namespace 'ThemedDropdown';
+
+@st-import Dropdown, [menu] from './dropdown/dropdown.st.css';
+@st-import Menu from './menu/menu.st.css';
+@st-import TextOption from './text-option/text-option.st.css';
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:5]: unused

File: src/components/themed-dropdown.st.css [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+.menu {
+    background-color: value(surface-color); /* to hide anything that is "behind" */
+    box-sizing: border-box; /* so that scrollbar will be the "inside" list */
+    box-shadow: 2px 2px 10px rgba(0, 0, 0, 0.3);
+    margin: 0;
+    padding: 0;
+    max-height: 8em;
+    overflow: auto; /* scrollbar */
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:43]: same comment from line 28

File: src/components/themed-dropdown.st.css [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+.themedDropdown::trigger {
+    border: 1px solid value(surface-color-secondary);
+    border-radius: 0.2em;
+
+    padding-left: 0.15em;
+}
+
+.themedDropdown::trigger::icon {
+    width: 1em;
+    height: 1em;
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:29]: should these values be in the themed stylesheet?
What is the distinction between themed and regular component stylesheet?

File: src/components/themed-dropdown.st.css [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+.themedDropdown {
+    -st-extends: Dropdown;
+}
+
+.themedDropdown::trigger {
+    border: 1px solid value(surface-color-secondary);
+    border-radius: 0.2em;
+
+    padding-left: 0.15em;
```
Comment (PeterShershov) [inline src/components/themed-dropdown.st.css:24]: same comment from line 28

File: src/components/dropdown/dropdown.tsx [hunk 19]
```diff
@@ -51,7 +69,7 @@ export const Dropdown: FC<{
  *          handleClose - closes the dropdown
  *          handleTrigger - triggers the dropdown open becomes closed and the other way around
  */
-function useToggle(openByDefault?: boolean): [boolean, () => void, () => void] {
+function useToggle(openByDefault: boolean): [boolean, () => void, () => void] {
```
Comment (PeterShershov) [inline src/components/dropdown/dropdown.tsx:72]: Not serious but I think that
```ts
function useToggle(openByDefault: boolean): { isOpen: boolean, close: () => void, toggle: () => void } {
```
is more readable

File: src/components/menu/menu.st.css [hunk 20]
```diff
@@ -0,0 +1,20 @@
...
+@st-import TextOption from "../text-option/text-option.st.css";
+@st-import Divider from "../divider/divider.st.css";
+
+.root {
+    -st-states: disabled;
+    position: absolute; /* so that it will overlay other items in the document */
+}
+
+.root:disabled{
+}
```
Comment (PeterShershov) [inline src/components/menu/menu.st.css:12]: no need for that

File: src/components/menu/menu.st.css [hunk 21]
```diff
@@ -0,0 +1,20 @@
...
+
+.textOption{
+    -st-extends: TextOption;
+}
+
+.divider{
```
Comment (PeterShershov) [inline src/components/menu/menu.st.css:18]: do you use stylable-intelligence? should prettify this with an extra space automatically

**Review by eladdo92 — APPROVED** [review]

**Pull request merged**